rfcs/45-stageless.md

```diff
@@ -0,0 +1,977 @@
+
+### What sugar should we use for adding multiple systems at once?
+
+Convenience methods like `add_systems` are important for reducing boilerplate. For these methods, we need to be able to refer to collections of systems and system sets.
+
+**In summary:**
+
+- builder syntax: marginal improvement over adding individual systems
+- array syntax: looks pretty but is literally impossible
+- tuple syntax: looks pretty but is limited to 12 elements
+- `vec!`-like macro syntax: looks OK but, eh, uses macros
+
+**Conclusion:** macro syntax for practicality, likely powered by builder syntax under the hood
+
+Below, we examine the the options by example.
+
+#### Builder pattern
+
+This was the strategy we used for `SystemSet`.
+
+```rust
+.add_systems(
+    Nodes::new()
+        .with(compute_attack), 
+        .with(compute_defense),
+        .with(check_for_crits),
+        .with(compute_damage),
+        .with(deal_damage),
+        .with(check_for_death)
+        .chain()
+        .in_set(GameSet::Combat)
+    )
+
+.add_systems(Nodes::new().with(run).with(jump).after(InputSet::ReadInput))
+```
+
+- looks bad
+  - not much better than adding things individually
+  - confusing placement of brackets
+  - no clear seperation between systems and the group config
+- no limit to number of elements
+
+#### Arrays
+
+```rust
+.add_systems([
+        compute_attack, 
+        compute_defense,
+        check_for_crits,
+        compute_damage,
+        deal_damage,
+        check_for_death,
+    ]
+    .chain()
+    .in_set(GameSet::Combat))
+
+.add_systems([run, jump].after(InputSet::ReadInput))
+```
+
+- looks pretty
+- no limit to number of elements
+- doesn't work (different functions are different types, arrays must be homogoneous)
+
+#### Tuples
+
+```rust
+.add_systems((
+        compute_attack, 
+        compute_defense,
+        check_for_crits,
+        compute_damage,
+        deal_damage,
+        check_for_death
+    )
+    .chain()
+    .in_set(GameSet::Combat)
+)
+
+.add_systems((run, jump).after(InputSet::ReadInput))
+```
+
+- looks pretty
+- limited to 12 elements
+- increases compile times
+- relies on type system magic
+  - requires invocation of the `all_tuples` macro to generate blanket impls
+  - (kinda cursed, but already used for system params, systems, and bundles)
+
+#### `vec!`-like macros
+
+This strategy is used for the `vec![1,2,3]` macro in the standard library.
+
+```rust
+.add_systems(systems![
+      compute_attack, 
+      compute_defense,
+      check_for_crits,
+      compute_damage,
+      deal_damage,
+      check_for_death
+   ]
+   // can also use `chain![...]` for slightly more compact shortcut
+   .chain() 
+   .in_set(GameSet::Combat)
+)
+
+.add_systems(systems![run, jump].after(InputSet::ReadInput))
+```
+
+- looks OK (some boilerplate)
+- no limit to number of elements
+- relies on regular, declarative macro magic
+
+## Future possibilities

```

Comment:
I think that future possibilities can be grouped in different categories, like API improvements and performance optimizations.